pkg/controller/windowsmachine/windowsmachine_controller.go

```diff
@@ -441,19 +439,45 @@ func (r *ReconcileWindowsMachine) isAllowedDeletion(machine *mapi.Machine) bool
 		return false
 	}
 
-	if minHealthyCount == *(windowsMachineSet.Spec.Replicas) {
+	// Allow deletion if there is only one machine in the Windows MachineSet
+	totalWindowsMachineCount := *windowsMachineSet.Spec.Replicas
+	if maxUnhealthyCount == totalWindowsMachineCount {
 		return true
 	}
 
 	totalHealthy := 0
 	for _, ma := range machines.Items {
-		// Machines are determined as Healthy when they are part of Windows MachineSet and are
-		// in the Running Status
-		if len(ma.OwnerReferences) != 0 && ma.OwnerReferences[0].Name == machinesetName {
-			if ma.Status.Phase != nil && *ma.Status.Phase == "Running" && ma.Status.NodeRef != nil {
-				totalHealthy += 1
-			}
+		// Increment the count if the machine is identified as healthy and is a part of given Windows MachineSet and
+		// on which deletion is not already initiated.
+		if len(machine.OwnerReferences) != 0 && ma.OwnerReferences[0].Name == machinesetName &&
+			r.isWindowsMachineHealthy(&ma) && ma.DeletionTimestamp.IsZero() {
+			totalHealthy += 1
 		}
 	}
-	return totalHealthy-1 >= minHealthyCount
+
+	return totalWindowsMachineCount-int32(totalHealthy) < maxUnhealthyCount
+}
+
+// isWindowsMachineHealthy determines if the given Machine object is healthy. A Windows machine is considered
+// unhealthy if -
+// 1. Machine is not in a 'Running' phase
+// 2. Machine is not associated with a Node object
+// 3. Associated Node object doesn't have a Version annotation
+func (r *ReconcileWindowsMachine) isWindowsMachineHealthy(machine *mapi.Machine) bool {
+	if (machine.Status.Phase == nil || *machine.Status.Phase != "Running") &&
+		machine.Status.NodeRef == nil {
+		return false
+	}
+
+	// Get node associated with the machine
+	node, err := r.k8sclientset.CoreV1().Nodes().Get(context.TODO(), machine.Status.NodeRef.Name, meta.GetOptions{})
+	if err != nil {
+		return false
+	}
+	_, present := node.Annotations[nodeconfig.VersionAnnotation]
+	if !present {
+		return false
```

Comment:
Should we be logging each case at INFO level? Otherwise there would be no feedback to the user.

Reply:
Not sure if logging every machine, every time a node is considered to be deleted will help. An unhealthy machine here states that the corresponding node is not fully configured and is being logged in previous states. 

I can add a INFO log [here](https://github.com/openshift/windows-machine-config-operator/pull/148/files/4cd2aaa823866f6bafef3ff0295a7326db48c946#diff-f6cf75653deff7ed72be94f059acc292d15113ed7c9f9ce0fdc5e842fa71aa72R458) stating the count of healthy machines in the desired machineset.

Reply:
>Not sure if logging every machine, every time a node is considered to be deleted will help.

I am asking for a log for the case when we decided **not** to delete a machine.

>An unhealthy machine here states that the corresponding node is not fully configured and is being logged in previous states.

This seems to imply the cases I am talking about is covered.

> I can add a INFO log here stating the count of healthy machines in the desired machineset.

+1